tests/Unit/Dashboard/ImportantMailWidgetTest.php

```diff
@@ -0,0 +1,128 @@
+
+	private IL10N&MockObject $l10n;
+	private IURLGenerator&MockObject $urlGenerator;
+	private IUserManager&MockObject $userManager;
+	private AccountService&MockObject $accountService;
+	private MailSearch&MockObject $mailSearch;
+	private IInitialState&MockObject $initialState;
+	private ImportantMailWidget $widget;
+
+	protected function setUp(): void {
+		parent::setUp();
+
+		$this->l10n = $this->createMock(IL10N::class);
+		$this->urlGenerator = $this->createMock(IURLGenerator::class);
+		$this->userManager = $this->createMock(IUserManager::class);
+		$this->accountService = $this->createMock(AccountService::class);
+		$this->mailSearch = $this->createMock(MailSearch::class);
+		$this->initialState = $this->createMock(IInitialState::class);
+
+		$this->widget = new ImportantMailWidget(
+			$this->l10n,
+			$this->urlGenerator,
+			$this->userManager,
+			$this->accountService,
+			$this->mailSearch,
+			$this->initialState,
+			'bob'
+		);
+	}
+
+	public function testGetItems(): void {
+		$user = $this->createMock(IUser::class);
+		$this->userManager->expects($this->once())
+			->method('get')
+			->willReturn($user);
+
+		$message1 = new Message();
+		$message1->setSubject('Important');
+		$message1->setMailboxId(1);
+
+		$message2 = new Message();
+		$message2->setSubject('Also important');
+		$message2->setMailboxId(2);
+
+		$this->mailSearch->expects($this->once())
+			->method('findMessagesGlobally')
+			->with($user, $this->callback(function (GlobalSearchQuery $query) {
+				self::assertCount(1, $query->getFlags());
+				self::assertNull($query->getStart());
+				return true;
+			}))
+			->willReturn([$message1, $message2]);
+
+		$mailAccount = new MailAccount();
+		$account = new Account($mailAccount);
+
+		$this->accountService->expects($this->once())
+			->method('findByUserId')
+			->willReturn([$account]);
+
+		$items = $this->widget->getItems('bob', null, 7);
+
+		$this->assertCount(2, $items);
+	}
+
+	public function testGetItemsWithSince(): void {
+		$user = $this->createMock(IUser::class);
+		$this->userManager->expects($this->once())
+			->method('get')
+			->willReturn($user);
+
+		$message1 = new Message();
+		$message1->setSubject('Important');
+		$message1->setMailboxId(1);
+
+		$message2 = new Message();
+		$message2->setSubject('Also important');
+		$message2->setMailboxId(2);
+
+		$this->mailSearch->expects($this->once())
+			->method('findMessagesGlobally')
+			->with($user, $this->callback(function (MailSearchQuery $query) {
+				self::assertCount(1, $query->getFlags());
+				self::assertNotNull($query->getStart());
+				return true;
+			}))
+			->willReturn([$message1, $message2]);
+
+		$mailAccount = new MailAccount();
+		$account = new Account($mailAccount);
+
+		$this->accountService->expects($this->once())
+			->method('findByUserId')
+			->willReturn([$account]);
+
+		$items = $this->widget->getItems('bob', '1745340000', 7);
+
+		$this->assertCount(2, $items);
+	}
```

Comment:
Tip for readable unit tests: structure them into three blocks: *arrange*, *act* and *assert*. Separate the blocks by one newline.

Reply:
You mentioned AAA a couple of times, yes 😉 I've changed the tests to follow the AAA pattern, even though I'm not a fan of it and find them rather difficult to read. Most of our tests have complex "Arrange" sections, and understanding them is quite hard when the code isn't at least somewhat structured.